integration/diff_test.go

```diff
@@ -0,0 +1,540 @@
+				return
+			}
+
+			require.Equal(t, math.Signbit(math.Copysign(0.0, +1)), math.Signbit(actual))
+		})
+	}
+}
+
+func TestDiffUpdateProduceInfinity(t *testing.T) {
+	t.Parallel()
+
+	ctx, collection := setup.Setup(t)
+	_, err := collection.InsertOne(ctx, bson.D{{"_id", "number"}, {"v", int32(42)}})
+	require.NoError(t, err)
+
+	_, err = collection.UpdateOne(ctx, bson.D{{"_id", "number"}}, bson.D{{"$mul", bson.D{{"v", math.MaxFloat64}}}})
+
+	if setup.IsMongoDB(t) {
+		require.NoError(t, err)
+		return
+	}
+
+	expected := mongo.CommandError{
+		Code: 2,
+		Name: "BadValue",
+		Message: `update produces invalid value: { "v": +Inf }` +
+			` (update operations that produce infinity values are not allowed)`,
+	}
+	AssertEqualCommandError(t, expected, err)
+}
+
+func TestDiffDocumentValidation(t *testing.T) {
+	t.Parallel()
+
+	ctx, collection := setup.Setup(t, shareddata.Scalars)
+
+	t.Run("Insert", func(t *testing.T) {
+		t.Parallel()
+
+		for name, tc := range map[string]struct { //nolint:vet // use only for testing
+			doc bson.D
+
+			err error
+		}{
+			"DollarSign": {
+				doc: bson.D{{"$foo", "bar"}},
+				err: mongo.WriteException{WriteErrors: []mongo.WriteError{{
+					Index:   0,
+					Code:    2,
+					Message: `invalid key: "$foo" (key must not start with '$' sign)`,
+				}}},
+			},
+			"DotSign": {
+				doc: bson.D{{"foo.bar", "baz"}},
+				err: mongo.WriteException{WriteErrors: []mongo.WriteError{{
+					Index:   0,
+					Code:    2,
+					Message: `invalid key: "foo.bar" (key must not contain '.' sign)`,
+				}}},
+			},
+			"Infinity": {
+				doc: bson.D{{"foo", math.Inf(1)}},
+				err: mongo.WriteException{WriteErrors: []mongo.WriteError{{
+					Code:    2,
+					Message: `invalid value: { "foo": +Inf } (infinity values are not allowed)`,
+				}}},
+			},
+			"NegativeInfinity": {
+				doc: bson.D{{"foo", math.Inf(-1)}},
+				err: mongo.WriteException{WriteErrors: []mongo.WriteError{{
+					Code:    2,
+					Message: `invalid value: { "foo": -Inf } (infinity values are not allowed)`,
+				}}},
+			},
+			"NaN": {
+				doc: bson.D{{"_id", "nan"}, {"foo", math.NaN()}},
+				err: mongo.CommandError{
+					Code: 2,
+					Name: "BadValue",
+					Message: `wire.OpMsg.Document: validation failed for { insert: "TestDiffDocumentValidation", ordered: true, ` +
+						`$db: "TestDiffDocumentValidation", documents: [ { _id: "nan", foo: nan.0 } ] } with: NaN is not supported`,
+				},
+			},
+		} {
+			name, tc := name, tc
+			t.Run(name, func(t *testing.T) {
+				t.Parallel()
+
+				_, err := collection.InsertOne(ctx, tc.doc)
+
+				if setup.IsMongoDB(t) {
+					require.NoError(t, err)
+					return
+				}
+
+				assert.Equal(t, tc.err, UnsetRaw(t, err))
+			})
+		}
+	})
+
+	t.Run("Update", func(t *testing.T) {
+		t.Parallel()
+
+		for name, tc := range map[string]struct {
+			doc bson.D
+			err mongo.WriteError
+		}{
+			"DotSign": {
+				doc: bson.D{{"$set", bson.D{{"foo", bson.D{{"bar.baz", "qaz"}}}}}},
+				err: mongo.WriteError{
```

Comment:
This was `CommandError` in dance
https://github.com/FerretDB/dance/blob/main/tests/diff/document_validation_test.go#L110

I assume this test was not passing in dance with new postgresql backend, because update returns WriteError on failed validation.